### PR TITLE
IOW-979 -- refactor to steam statistic time series WaterML:

### PR DESCRIPTION
**Pull request by orosz-usgs**
*) Refactored sql to select rows
  *) Updated result handler to accommodate time series data.
  *) Xml refactored to comply with WaterML schema.
  *) Added validation of WaterML schema in test.

Before making a pull request
----------------------------

- [ ] Make sure all tests run
- [ ] Update the changelog appropriately
- [ ] Make sure the Swagger docs are accurate

Title
-----------
Brief description of changes. Reference the JIRA ticket if appropriate

Description
-----------
If no JIRA ticket is referenced, describe the changes made. Note anything that you want the reviewers to know while
reviewing your pull request

After making a pull request
---------------------------
- [ ] If appropriate, put the link to the PR in the JIRA ticket
- [ ] Assign someone to review unless the change is trivial

### CODE REVIEW

**Review by duselmann — COMMENTED** [review]
few comments

File: src/main/java/gov/usgs/wma/waterdata/format/WaterMLPointToXmlResultHandler.java [hunk 1]
```diff
@@ -114,69 +125,113 @@ private void addHeaderMetData() throws XMLStreamException {
...
         startElement(ObsAndMeasure.NAMESPACE, "OM_Observation");
+        addObservationTimes(point);
         startElement(ObsAndMeasure.NAMESPACE, "procedure");
         startElement(ObsAndMeasure.NAMESPACE, "ObservationProcess");
-        startElement(ObsAndMeasure.NAMESPACE, "parameter");
-        writer.writeAttribute(XLINK_NS, "title", "Statistic");
-        writer.writeAttribute(XLINK_NS, "href",
-                "http://waterdata.usgs.gov/nwisweb/rdf?statCd=00011\"");
-        endElement("parameter");
+        startElement("parameter");
```
[the remark this answers is not in the document]
Comment (orosz-usgs) [reply src/main/java/gov/usgs/wma/waterdata/format/WaterMLPointToXmlResultHandler.java:139]: sure, can do.

File: src/main/java/gov/usgs/wma/waterdata/format/WaterMLPointToXmlResultHandler.java [hunk 2]
```diff
@@ -114,69 +125,113 @@ private void addHeaderMetData() throws XMLStreamException {
...
+            endElement("TimePeriod");
+            endElement("phenomenonTime");
+        }
+        if(point.getPhenomenonTimeEnd() != null) {
+            startElement(ObsAndMeasure.NAMESPACE, "resultTime");
+            startElement(GeoMarkupLanguage.NAMESPACE, "TimeInstant");
+            addElement(GeoMarkupLanguage.NAMESPACE, "timePosition", point.getPhenomenonTimeEnd());
+            endElement("TimeInstant");
+            endElement("resultTime");
+        }
```
[the remark this answers is not in the document]
Comment (orosz-usgs) [reply src/main/java/gov/usgs/wma/waterdata/format/WaterMLPointToXmlResultHandler.java:173]: The "timePosition" element should always be populated. nothing really dynamic.  In discrete data, it is currently not.
The if blocks were added to ensure the top level phenomenonTime would not be rendered if there were no
dates defined.

File: src/main/java/gov/usgs/wma/waterdata/format/WaterMLPointToXmlResultHandler.java [hunk 3]
```diff
@@ -114,69 +125,113 @@ private void addHeaderMetData() throws XMLStreamException {
...
+            if(point.getInterpolationTypeRef() != null) {
+                writer.writeAttribute("href", point.getInterpolationTypeRef());
+            }
+            if(point.getInterpolationTypeDesc() != null) {
+                writer.writeAttribute("title", point.getInterpolationTypeDesc());
+            }
+            endElement("interpolationType");
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (orosz-usgs) [reply src/main/java/gov/usgs/wma/waterdata/format/WaterMLPointToXmlResultHandler.java:234]: There might be a way to have addElement() handle this, with new optional parameter. Think that should be a recfactor for
 another time.
Wrote a new handle to help eliminate part of the if  != null statements.

File: src/main/java/gov/usgs/wma/waterdata/timeseries/TimeSeriesController.java [hunk 4]
```diff
@@ -80,16 +83,24 @@ public void getTimeSeries(@PathVariable(value = PARAM_COLLECTION_ID) String coll
 			@RequestParam(value = "f", required = false, defaultValue = "json")
 			String mimeType,
 			HttpServletResponse response)
-					throws HttpMediaTypeNotAcceptableException, IOException {
+			throws HttpMediaTypeNotAcceptableException, IOException, XMLStreamException {
 
 		ContentType contentType = determineContentType(mimeType);
 		String rtn = null;
+		boolean streamingOutput = false;
+		boolean outputStreamed = false;
```
[the remark this answers is not in the document]
Comment (orosz-usgs) [reply src/main/java/gov/usgs/wma/waterdata/timeseries/TimeSeriesController.java:91]: How about changing 'outputStreamed' to 'responseWritten'?

File: src/main/java/gov/usgs/wma/waterdata/timeseries/TimeSeriesController.java [hunk 5]
```diff
@@ -98,6 +109,12 @@ public void getTimeSeries(@PathVariable(value = PARAM_COLLECTION_ID) String coll
 			rtn = ogc404Payload;
 		}
 
-		response.getWriter().print(rtn);
+		if (!outputStreamed) {
+			if (streamingOutput) {
+				response.getOutputStream().print(rtn);
```
[the remark this answers is not in the document]
Comment (orosz-usgs) [reply src/main/java/gov/usgs/wma/waterdata/timeseries/TimeSeriesController.java:114]: Yes, it seems like there should be a some helper method or class that could eliminate this if block.
I'll try it out...

File: src/main/java/gov/usgs/wma/waterdata/timeseries/TimeSeriesController.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (duselmann) [reply src/main/java/gov/usgs/wma/waterdata/timeseries/TimeSeriesController.java:91]: You know, that seem clearer to me, and was not what I thought that variable meant. Thanks.

File: src/main/resources/mybatis/discreteGWWaterMLPoint.xml [hunk 6]
```diff
@@ -2,31 +2,29 @@
...
+				|| '/items/' || ml.monitoring_location_identifier || '?f=json' as "monLocReference",
+			site_name "siteName",
 			date_measured_raw "resultDateTimeUTC", /* timestamp of the UTC time */
 			display_result "resultValue",
 			parm.parm_unt_tx "resultUnit",
 			parameter_code "pcode",
-		    parm.parm_ds "pcodeDesc",
+			parm.parm_ds "pcodeDesc",
+			'Daily random instantaneous values' "statisticDesc",
+			'http://waterdata.usgs.gov/nwisweb/rdf?statCd=00011' "statisticReference",
```
[the remark this answers is not in the document]
Comment (orosz-usgs) [reply src/main/resources/mybatis/discreteGWWaterMLPoint.xml:18]: Just to leave for another refactor to have this added to the parameter map and not be hard coded.
Did not make the cut off for this ticket.

File: src/main/resources/mybatis/discreteGWWaterMLPoint.xml [hunk 7]
```diff
@@ -2,31 +2,29 @@
...
-		<result property="qualifiersAsJson" column="qualifiersAsJson"/>
-		<result property="pcodeDesc" column="pcodeDesc"/>
-		<result property="status" column="status"/>
-	</resultMap>
 
-	<select id="getDiscreteGWMLPoint" parameterType="map" resultMap="pointResult" fetchSize="500">
+	<select id="getDiscreteGWMLPoint" parameterType="map" resultMap="groundwaterML.pointResult" fetchSize="500">
 		select
-			monitoring_location_identifier "monLocIdentifier",
+			dgw.monitoring_location_identifier "monLocIdentifier",
```
[the remark this answers is not in the document]
Comment (orosz-usgs) [reply src/main/resources/mybatis/discreteGWWaterMLPoint.xml:8]: not sure exact what you mean,  maybe have a quick call.

[the remark this answers is not in the document]
Comment (duselmann) [reply src/main/resources/mybatis/discreteGWWaterMLPoint.xml:8]: "monLocIdentifier" vs 'mld" vs "monitoring_location_identifier"
This query uses the alias "monLocIdentifier" and was wondering if we could standardize on a short form, "MLID" (no matter the case). Not a big deal and I am ok with a quick call if this did not clear it up.

File: src/main/resources/mybatis/discreteGWWaterMLPoint.xml (hunk 6 above)
[the remark this answers is not in the document]
Comment (duselmann) [reply src/main/resources/mybatis/discreteGWWaterMLPoint.xml:18]: sure.

File: src/main/java/gov/usgs/wma/waterdata/format/WaterMLPointToXmlResultHandler.java [hunk 8]
```diff
@@ -255,6 +324,22 @@ private void addElement(String name, String value) throws XMLStreamException {
         }
     }
 
+    private void addElement(String namespaceURI, String name, String value) throws XMLStreamException {
+        if (value == null) {
+            writer.writeEmptyElement(namespaceURI, name);
```
[the remark this answers is not in the document]
Comment (orosz-usgs) [reply src/main/java/gov/usgs/wma/waterdata/format/WaterMLPointToXmlResultHandler.java:329]: yes, writeEmptyElement() does that.

File: src/main/resources/mybatis/discreteGWWaterMLPoint.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (orosz-usgs) [reply src/main/resources/mybatis/discreteGWWaterMLPoint.xml:8]: got it. However, I believe it should be 'featureId'. That is the term used
throughout the rest of the observations service, in the Swagger doc, Java controllers,
MyBatis parameter map...

File: src/main/java/gov/usgs/wma/waterdata/format/WaterMLPointToXmlResultHandler.java [hunk 9]
```diff
@@ -269,10 +354,24 @@ private void endElement(String name) throws XMLStreamException {
               // convert to csv separated
               String csv =  point.getQualifiersAsJson().replace("[", "").
                       replace("]","").replace("\"", "");
-              qualifierList =  Arrays.asList(csv.split(",", 10));
+              qualifierList =  Arrays.asList(csv.split(",", 0));
```
[the remark this answers is not in the document]
Comment (orosz-usgs) [reply src/main/java/gov/usgs/wma/waterdata/format/WaterMLPointToXmlResultHandler.java:357]: yes, split(regex) actually just calls  split(regex, 0)

**Review by duselmann — APPROVED** [review]

File: src/main/resources/mybatis/discreteGWWaterMLPoint.xml (hunk 7 above)
[the remark this answers is not in the document]
Comment (duselmann) [reply src/main/resources/mybatis/discreteGWWaterMLPoint.xml]: yes, that makes sense. We should use featureId for any public API.

**Pull request merged**